### PR TITLE
Account for conditional mods in assignment algorithm

### LINKED ISSUE #7198: 'Show Mod Placement' places some mods in less than optimal places
### DIM Version

Version 6.88.1.1000808 (beta), built on 10/29/2021, 8:40:00 AM

### Browser Details

Version 96.0.4664.18 (Official Build) beta (64-bit)

### OS Details

Windows 10

### Describe the bug

I have setup a loadout on my Hunter with 'Powerful Friends' specifically on my arc helmet which is a raid helmet and it has the extra modslot for slotting an arc mod with 1 cost.  This is so that I can get the +20 mobility wihout using another arc charged with light mod on another armor piece.  When I click the 'Show Mod Placement', DIM randomly assigns the mods to my armor pieces but puts the 'Powerful Friends' mod on the leg armor which doesn't have the extra raid mod slot.
2021-10-29 16_39_12-DIM

Also another issue is that the Mobility stat is showing as 142 which is way too high.  It should be 102.

### Logs

_No response_

### PR DESCRIPTION
**Pull request by ryan-rushton**
This accounts for conditional mods in the mod assignment algorithm used for displaying mods and figuring out which mods we are going to socket into which item on loadout application.

Closes #7198

### CODE REVIEW

File: src/app/loadout/mod-assignment-utils.ts [hunk 1]
```diff
@@ -563,6 +599,53 @@ function calculateEnergyChange(
...
+  }
+
+  return activeMods;
+}
+
+/**
+ * Determines whether a conditional mod has had its requirements met by the other mods.
+ * Currently only used for radiant light and powerful friends
+ */
+function isPlugActive(
```
Comment (bhollis) [inline src/app/loadout/mod-assignment-utils.ts:625]: Didn't we have this somewhere else to determine whether the conditional stats were active on these?

**Review by bhollis — APPROVED** [review]
I don't really understand how this works - are we using the totalActiveConditionalMods somewhere as a primary sorting factor?

File: src/app/loadout/mod-assignment-utils.ts [hunk 2]
```diff
@@ -106,8 +110,13 @@ export function fitMostMods(
   const bucketSpecificAssignments: ModAssignments = {};
 
   // just an arbitrarily large number
+  // The total cost to assign all the mods to a set
+  // this includes energy used to upgrade and energy wasted on changing elements
   let assignmentEnergyCost = 10000;
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout/mod-assignment-utils.ts:115]: No, I will do that. I didn't actually know it was a thing.

File: src/app/loadout/mod-assignment-utils.ts [hunk 3]
```diff
@@ -231,21 +240,39 @@ export function fitMostMods(
...
-          (unassignedModCount <= assignmentUnassignedModCount &&
-            energyUsedAndWasted < assignmentEnergyCost)
-        ) {
+        // Skip further checks if energy used and wasted is higher for this assignment
+        if (energyUsedAndWasted > assignmentEnergyCost) {
+          continue;
+        }
+
+        let totalActiveConditionalMods = 0;
+        const allAssignedMods = Object.values(assignments).flatMap(
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout/mod-assignment-utils.ts:259]: Can do.

File: src/app/loadout/mod-assignment-utils.ts (hunk 1 above)
Comment (ryan-rushton) [reply src/app/loadout/mod-assignment-utils.ts:625]: Yes, I basically took that function and pulled the parts I needed out of it. I nearly added this to be right along side it, but I only need it in this one place, and I don't currently see a use case for it outside of this mod algorithm.

**Comment by ryan-rushton** [comment]
> I don't really understand how this works - are we using the totalActiveConditionalMods somewhere as a primary sorting factor?

Not quite. The mod assignment algorithm is very similar to the main loadout loop with the one main exception being we only keep a single assignment. So it works in the following manner,
1. Create all permutations of the slot independent mods
2. Loop over all those permutations
3. For each item in a consistent order (helmet -> class item), take the items index position and get the mod out of each slot independent permutations using it (i.e. `combatMods[i]`, `generalMods[i]`, ect). 
4. Validate the mod energies are compatible, i.e. cant have 1 arc and 1 solar on an item, must add up to less than 10 with the slot specific mods
5. If all mod energies are valid for all items, we then make a bunch of calculations to see if this assignment is better than the currently saved one, these calculations are now
  - the number of unassigned mods
  - the energy investment needed to apply all the mods, this has a energy wasted and energy used to upgrade method
  - are we activating more conditional mods **<-- This PR adds this**

**Review by nev-r — APPROVED** [review]

**Comment by bhollis** [comment]
So should I expect that the effect of this is that if all things are equal it'll *prefer* to put conditional mods together with the things that activate them, but it'll still leave them inactive if it has to to fit all the mods in?

**Comment by ryan-rushton** [comment]
> So should I expect that the effect of this is that if all things are equal it'll prefer to put conditional mods together with the things that activate them, but it'll still leave them inactive if it has to to fit all the mods in?

Yes that is the intent, so what it will do when checking a new assignment is
- If there are less unassigned mods, save new assignment as best, else
- if there are more conditional mods being activated, save new assignment as best, else
- if less energy is used/wasted, save new assignment as best

**Pull request merged**